### PR TITLE
Allow object downloads without authentication from certain remote object stores

### PR DESCRIPTION
**Pull request by sadhansood**
## Description 

We want to allow object download without authentication from certain remote stores with free egress i.e. cloudflare, wasabi, etc. This PR adds a new downloader module which allows us to do that for S3, GCS, etc
Next step would be to integrate it in sui tool and let users benefit from it.

## Test Plan 

Added unit test. Also tested with S3 and found it to be working

### CODE REVIEW

File: crates/sui-storage/src/object_store/downloader/gcs.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        let client = builder.https_only(false).build()?;
+        let bucket_name_encoded = percent_encode(bucket.as_bytes(), NON_ALPHANUMERIC).to_string();
+
+        Ok(Self {
+            client,
+            bucket_name_encoded,
+        })
+    }
+
+    async fn get(&self, path: &Path) -> Result<GetResult> {
```
Comment (johnjmartin) [inline crates/sui-storage/src/object_store/downloader/gcs.rs:34]: assuming we enable it, both gcs and s3 http endpoints support listing the contents of the bucket, so could add support for that later. (it's not supported in cloudflare r2)

File: crates/sui-storage/src/object_store/downloader/mod.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    let response = request.send().await.context("failed to get")?;
+    let meta = header_meta(location, response.headers()).context("Failed to get header")?;
+    let stream = response
+        .bytes_stream()
+        .map_err(|source| Error::Generic {
+            store,
+            source: Box::new(source),
+        })
+        .boxed();
+    Ok(GetResult {
```
Comment (johnjmartin) [inline crates/sui-storage/src/object_store/downloader/mod.rs:81]: This is cool 👍 , makes it easily pluggable with parts of the code that rely on `object_store`?

**Review by johnjmartin — APPROVED** [review]

**Review by williampsmith — APPROVED** [review]

File: crates/sui-storage/src/object_store/downloader/gcs.rs (hunk 1 above)
Comment (sadhansood) [reply crates/sui-storage/src/object_store/downloader/gcs.rs:34]: yeah definitely, we should be able to add list support as well. I looked at it, and seemed a little complicated to left it as a todo for later

File: crates/sui-storage/src/object_store/downloader/mod.rs (hunk 2 above)
Comment (sadhansood) [reply crates/sui-storage/src/object_store/downloader/mod.rs:81]: yes exactly, that is the intent. I think there is still a little more work to make retryable `get` util calls work with passowrdless downloader but should not be too hard

**Pull request merged**